src/IO/RemoteReadBufferCache.h

```diff
@@ -0,0 +1,232 @@
+            const std::string & path_,
+            UInt64 last_modification_timestamp_,
+            size_t file_size_):
+        schema(schema_),
+        cluster(cluster_),
+        path(path_),
+        last_modification_timestamp(last_modification_timestamp_),
+        file_size(file_size_)
+    {}
+
+    std::string schema; // Hive, S2 etc.
+    std::string cluster;
+    std::string path;
+    UInt64 last_modification_timestamp;
+    size_t file_size;
+};
+
+/**
+ *
+ */
+class RemoteCacheController
+{
+public:
+    RemoteCacheController(
+        const RemoteFileMeta &meta,
+        const std::filesystem::path & local_path_,
+        size_t cache_bytes_before_flush_,
+        std::shared_ptr<ReadBuffer> readbuffer_,
+        std::function<void(RemoteCacheController *)> const & finish_callback);
+    ~RemoteCacheController(); // the local files will be deleted in descontructor
+
+    // recover from local disk
+    static std::shared_ptr<RemoteCacheController>
+    recover(const std::filesystem::path & local_path, std::function<void(RemoteCacheController *)> const & finish_callback);
+
+    /**
+     * Called by LocalCachedFileReader, must be used in pair
+     * The second value of the return tuple is the local_path to store file.
+     * It will be empty if the file has not been downloaded
+     */
+    std::tuple<FILE *, std::filesystem::path> allocFile();
+    void deallocFile(FILE * fs_);
+
+    /**
+     * when allocFile be called, count++. deallocFile be called, count--.
+     * the local file could be deleted only count==0
+     */
+    inline bool closable()
+    {
+        std::lock_guard lock{mutex};
+        return opened_file_streams.empty() && remote_readbuffer == nullptr;
+    }
+    void close();
+
+    /**
+     * called in LocalCachedFileReader read(), the reading process would be blocked until
+     * enough data be downloaded.
+     * If the file has finished download, the process would be block
+     */
+    RemoteReadBufferCacheError waitMoreData(size_t start_offset_, size_t end_offset_);
+
+    inline size_t size() const { return current_offset; }
+
+    inline const std::filesystem::path & getLocalPath() { return local_path; }
+    inline const std::string & getRemotePath() { return remote_path; }
+
+    inline UInt64 getLastModificationTimestamp() const { return last_modification_timestamp; }
+    inline void markInvalid()
+    {
+        std::lock_guard lock(mutex);
+        valid = false;
+    }
+    inline bool isValid()
+    {
+        std::lock_guard lock(mutex);
+        return valid;
+    }
+
+private:
+    // flush file and meta info into disk
+    void flush(bool need_flush_meta_ = false);
+
+    void backgroupDownload(std::function<void(RemoteCacheController *)> const & finish_callback);
+
+    std::mutex mutex;
+    std::condition_variable more_data_signal;
+
+    std::set<FILE *> opened_file_streams;
+
+    // meta info
+    bool download_finished;
+    bool valid;
+    size_t current_offset;
```

Comment:
looks like you have races around `current_offset` and it is not even atomic, though I do not think that making it atomic is a good solution.

Reply:
current_offset is accessed in the lock protection